server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
-        int change = m.getPages().size() - driver.getPages().size();
-        if (log.isLoggable(Level.FINE)) {
-          log.fine(
-              "ApplicationSentListingMessage: message pages: " + m.getPages().size() + ", change: "
-              + change);
-        }
+  private void processPagesForSafariApp(ApplicationSentListingMessage applicationSentListingMessage) {
+    boolean isPagesAdded = isPagesAdded(applicationSentListingMessage);
+    if (log.isLoggable(Level.FINE)) {
+      log.fine("ApplicationSentListingMessage: message pages count: " + applicationSentListingMessage.getPages().size()
+          + ", pages added?: " + isPagesAdded);
+    }
 
-        // a new page appeared. The driver needs to know.
-        if (change > 0) {
-          List<WebkitPage> pages = new ArrayList<WebkitPage>();
-          // keep all the pages currently known to the driver
-          pages.addAll(driver.getPages());
+    // A new page appeared. The driver needs to know.
+    if (isPagesAdded) {
+      processAddedPages(applicationSentListingMessage);
+    } else {
+      processRemovedPages(applicationSentListingMessage);
+    }
+    sync.signalSimSentPages();
+  }
 
-          // remove all the pages we already know of from the message
-          for (WebkitPage p : driver.getPages()) {
-            m.getPages().remove(p);
-          }
+  private boolean arePagesNotChanged(ApplicationSentListingMessage applicationSentListingMessage) {
+    return WebkitPage.equals(applicationSentListingMessage.getPages(), driver.getPages());
+  }
 
-          if (m.getPages().size() == 0) {
-            throw new WebDriverException(m.getPages().size() + " new pages.");
-          }
+  private boolean isPagesAdded(ApplicationSentListingMessage applicationSentListingMessage) {
+    return applicationSentListingMessage.getPages().size() - driver.getPages().size() > 0;
+  }
 
-          // TODO there can be more than one 'new' UIWebView, picking the max one for now.
-          WebkitPage newOne = Collections.max(m.getPages());
-          pages.addAll(m.getPages());
-
-          driver.setPages(pages);
-
-          if (driver.getPages().size() == 0) {
-            //log.fine("first page. Nothing to do.");
-          } else if (newOne.isITunesAd()) {
-            //log.fine("itunes ad - ignoring it.");
-          } else {
-            WebkitPage focus = newOne;
-
-            if (session != null) {
-              waitForWindowSwitchingAnimation();
-              driver.switchTo(focus);
-            } else {
-              driver.switchTo(focus);
-            }
-          }
-          // a page disappeared, the driver needs to know.
-        } else if (change < 0) {
-          List<WebkitPage> old = new ArrayList<WebkitPage>();
-          old.addAll(driver.getPages());
+  private void processAddedPages(ApplicationSentListingMessage applicationSentListingMessage) {
+    List<WebkitPage> pages = new ArrayList<WebkitPage>();
```

Comment:
@krmahadevan 
Yes thats a point!! I just carried on the old code. I will change to the parameterized constructor;.